[New Concept]: Recursion (#3051)
exercism/python: concepts/recursion/about.md

# About

Recursion is a way to repeatedly execute code inside a function through the function calling itself.
It can be viewed as another way to loop/iterate. 
Like looping, a Boolean expression or `True/False` test is used to know when to stop the recursive execution.
_Unlike_ looping, recursion without termination in Python cannot not run infinitely.

> Values used in each function call are placed in their own frame on the Python interpreter stack.
> If the total amount of function calls takes up more space than the stack has room for, it will result in an error.
> 
> Python's interpreter stack is backed up by a C stack that is connected to the underlying operating system.
> Function calls that take up more memory space than the C stack has will cause a stack overflow.
> A stack overflow can crash Python or even the host operating system, if it is severe enough.
> To avoid this situation, the Python interpreter has a maximum stack depth that limits the number of frames that can be placed on its stack.
> The default is typically 999 frames, but varies by system.
> The recursion limit for a given version of Python can be found by calling [`sys.getrecursionlimit()`][getrecursionlimit]

Reviewer: ```suggestion
```

Reviewer: I like your phrasing below much better.  😄   This all can go in favor of that.

Author: Thanks, I think I'd like to stick with what I wrote for this one.

                                     + [date(year, month, day_num + (ordinal - 1) * 7)])
    return paydates_for_year_rec(year, weekday, ordinal, month, day_num + 1, output)

    # find the second Wednesday of the month for all the months in 2022
    print(paydates_for_year_rec(2022, 3, 2, 1, 1, []))
    
```

Adya is happy that there are no more nested loops, no mutated state, and 2 fewer lines of code than her looping solution.
Later, when she learns about tuples,  Adya may consider other approaches, such as using a list comprehension with `Calendar.itermonthdates`.
She may also become concerned that the recursive approach uses more steps than the looping approach, and so is less performant.
But, when faced with a solution that involves a lot of nested looping, extensive state mutation, and complex conditional logic, Adya may try to work it out recursively first.
She may find it easier to initially reason through the problem when nesting, mutation, and complexity are minimized.
After she's worked out the basic logic she can then work at optimizing the recursive solution into a more performant looping approach.

> 

Reviewer: I like this example, but it feels in tone to me more like a concept exercise rather than an illustration of loops vs recursion.  My though here it that maybe we use the `collatz conjecture` example for code on  `loops` vs `recursion`, and we save **this** narrative for a recursion _concept exercise_.  

I don't think we lose anything by using `Collatz` here -- students will likely have tackled  `Collatz` when they do the numbers concept.  And that might work even better since they will be familiar with the code, and might then be able to reason about the differences better.

Another option is to use the code from this example, but remove/alter the narrative style.  I think the narrative is _good_ -- but its story-telling doesn't exactly mesh (for me) with the rest of the doc, which reads more "friendly/casual college lecture".

Author: I'm not sure collatz is as good an example for "why recursion", as it doesn't have the nested looping. But if the purpose is to just show the difference between the two outcomes, as opposed to the difference in the two thought processes, then I suppose collatz would be a straightforward, though somewhat dull, comparison.

I'm also not sure that the narrative style _needs_ to "mesh" in tone with the surrounding doc, as a story can serve a pedagogic purpose of drawing the student in to the recursive thought process in an engaging way. But if the story is off-putting, then yeah, it should go.

# About

Recursion is a way to repeatedly execute code inside a function through the function calling itself.
Functions that call themselves are know as _recursive_ functions.
Recursion can be viewed as another way to loop/iterate. 
And like looping, a Boolean expression or `True/False` test is used to know when to stop the recursive execution.
_Unlike_ looping, recursion without termination in Python cannot not run infinitely.
Values used in each function call are placed in their own frame on the Python interpreter stack.
If the total amount of function calls takes up more space than the stack has room for, it will result in an error.

## Why Recursion?

If there is risk of causing a stack error or overflow, why would anyone use a recursive strategy to solve a problem?
There may be situations where a solution is more readable and/or easier to reason through when expressed through recursion than when expressed through looping.
There may also be program constraints with using/mutating data, managing complexity, delegating responsibility, or organizing workloads. 

> 

Reviewer: ```suggestion

## Why Recursion? (Indira's Insecurity)

Indira has her monthly social security auto-deposited in her bank account on the second Wednesday of every month.
Indira is concerned about balancing her check book.
She is afraid she will write checks before her money is deposited.
She asks her granddaughter Adya to give her a list of dates her money will appear in her account.
Adya, who is just learning how to program in Python, writes a program based on her first thoughts.
She wants to return a list of the dates so it can be printed.
She decides to use the `date` class imported from `datetime`.
She wants to write a function that will work for any year.
In case the schedule changes, or in case other relatives want Adya to calculate the schedule for their deposits, she decides the function needs a parameter for the weekday. 
Finally, Adya decides that the function needs a parameter for which weekday of the month it is: the first, second, etc.
Putting all of that together, Adya comes up with this.

```python
from datetime import date

def paydates_for_year(year, weekday, ordinal):
    """Returns a list of the matching weekday dates.
    
    Keyword arguments:
    year    -- the year, e.g. 2022
    weekday -- the weekday, e.g. 3 (for Wednesday)
    ordinal -- which weekday of the month, e.g. 2 (for the second)
    """
    output = []

    for month in range(1, 13):
        for day_num in range(1, 8):
            if date(year, month, day_num).isoweekday() == weekday:
                output.append(date(year, month, day_num + (ordinal - 1) * 7))
                break
    return output

# find the second Wednesday of the month for all the months in 2022
print(paydates_for_year(2022, 3, 2))

```

The program works, but Adya wonders if she can refactor the code to use fewer lines with less nested looping.
She would also like to use less mutating of variables, such as `output`, `month`, and `day_num`, because she's read that it is good to minimize mutating state.
She's heard about recursion and thinks about how to change her program to use a recursive approach.
The variables that are created and mutated in her looping function will now be passed in as arguments.
Instead of mutating the variables inside her function, she will pass updated values as arguments to the next call of her function.
With those intentions she arrives at this recursive approach

```python
from datetime import date

def paydates_for_year_rec(year, weekday, ordinal, month, day_num, output):
    """Returns a list of the matching weekday dates
    
    Keyword arguments:
    year    -- the year, e.g. 2022
    weekday -- the weekday, e.g. 3 (for Wednesday)
    ordinal -- which weekday of the month, e.g. 2 (for the second)
    month   -- the month currently being processed
    day_num -- the day of the month currently being processed
    output  -- the list to be returned
    """
    if month == 13:
        return output
    if date(year, month, day_num).isoweekday() == weekday:
        return paydates_for_year_rec(year, weekday, ordinal, month + 1, 1, output
                                     + [date(year, month, day_num + (ordinal - 1) * 7)])
    return paydates_for_year_rec(year, weekday, ordinal, month, day_num + 1, output)

    # find the second Wednesday of the month for all the months in 2022
    print(paydates_for_year_rec(2022, 3, 2, 1, 1, []))
    
```

Adya is happy that there are no more nested loops, no mutated state, and 2 fewer lines of code than her looping solution.
Later, when she learns about tuples,  Adya may consider other approaches, such as using a list comprehension with `Calendar.itermonthdates`.
She may also become concerned that the recursive approach uses more steps than the looping approach, and so is less performant.
But, when faced with a solution that involves a lot of nested looping, extensive state mutation, and complex conditional logic, Adya may try to work it out recursively first.
She may find it easier to initially reason through the problem when nesting, mutation, and complexity are minimized.
After she's worked out the basic logic she can then work at optimizing the recursive solution into a more performant looping approach.
```

## Why Recursion? (Indira's Insecurity)

Indira has her monthly social security auto-deposited in her bank account on the second Wednesday of every month.
Indira is concerned about balancing her check book.
She is afraid she will write checks before her money is deposited.
She asks her granddaughter Adya to give her a list of dates her money will appear in her account.
Adya, who is just learning how to program in Python, writes a program based on her first thoughts.
She wants to return a list of the dates so it can be printed.
She decides to use the `date` class imported from `datetime`.
She wants to write a function that will work for any year.
In case the schedule changes, or in case other relatives want Adya to calculate the schedule for their deposits, she decides the function needs a parameter for the weekday. 
Finally, Adya decides that the function needs a parameter for which weekday of the month it is: the first, second, etc.
Putting all of that together, Adya comes up with this.

```python
from datetime import date

def paydates_for_year(year, weekday, ordinal):
    """Returns a list of the matching weekday dates.
    
    Keyword arguments:

>     year    -- the year, e.g. 2022
>     weekday -- the weekday, e.g. 3 (for Wednesday)
>     ordinal -- which weekday of the month, e.g. 2 (for the second)
>     """
>     output = []
> 
>     for month in range(1, 13):
>         for day_num in range(1, 8):
>             if date(year, month, day_num).isoweekday() == weekday:
>                 output.append(date(year, month, day_num + (ordinal - 1) * 7))
>                 break
>     return output
> 
> # find the second Wednesday of the month for all the months in 2022
> print(paydates_for_year(2022, 3, 2))
> 
> ```
> 
> The program works, but Adya wonders if she can refactor the code to use fewer lines with less nested looping.
> She would also like to use less mutating of variables, such as `output`, `month`, and `day_num`, because she's read that it is good to minimize mutating state.
> She's heard about recursion and thinks about how to change her program to use a recursive approach.
> The variables that are created and mutated in her looping function will now be passed in as arguments.
> Instead of mutating the variables inside her function, she will pass updated values as arguments to the next call of her function.
> With those intentions she arrives at this recursive approach
> 
> ```python
> from datetime import date
> 
> 
> 
> def paydates_for_year_rec(year, weekday, ordinal, month, day_num, output):
>     """Returns a list of the matching weekday dates
>     
>     Keyword arguments:
>     year    -- the year, e.g. 2022
>     weekday -- the weekday, e.g. 3 (for Wednesday)
>     ordinal -- which weekday of the month, e.g. 2 (for the second)
>     month   -- the month currently being processed
>     day_num -- the day of the month currently being processed
>     output  -- the list to be returned
>     """
>     if month == 13:
>         return output
>     if date(year, month, day_num).isoweekday() == weekday:
>         return paydates_for_year_rec(year, weekday, ordinal, month + 1, 1, output
>                                      + [date(year, month, day_num + (ordinal - 1) * 7)])
>     return paydates_for_year_rec(year, weekday, ordinal, month, day_num + 1, output)
> 
>     # find the second Wednesday of the month for all the months in 2022
>     print(paydates_for_year_rec(2022, 3, 2, 1, 1, []))
>     
> ```
> 
> Adya is happy that there are no more nested loops, no mutated state, and 2 fewer lines of code than her looping solution.
> Later, when she learns about tuples,  Adya may consider other approaches, such as using a list comprehension with `Calendar.itermonthdates`.
> She may also become concerned that the recursive approach uses more steps than the looping approach, and so is less performant.
> But, when faced with a solution that involves a lot of nested looping, extensive state mutation, and complex conditional logic, Adya may try to work it out recursively first.
> She may find it easier to initially reason through the problem when nesting, mutation, and complexity are minimized.
> After she's worked out the basic logic she can then work at optimizing the recursive solution into a more performant looping approach.
> 

Reviewer: ```suggestion

```

# About

Recursion is a way to repeatedly execute code inside a function through the function calling itself.
Functions that call themselves are know as _recursive_ functions.
Recursion can be viewed as another way to loop/iterate. 
And like looping, a Boolean expression or `True/False` test is used to know when to stop the recursive execution.
_Unlike_ looping, recursion without termination in Python cannot not run infinitely.
Values used in each function call are placed in their own frame on the Python interpreter stack.
If the total amount of function calls takes up more space than the stack has room for, it will result in an error.

> 

Reviewer: _Here is where I would insert a write-up on "#7" -- how looping and recursion "feel" like they execute the same way, but they look very different when you compare steps, calls, and how each uses stack frames._

# About

Recursion is a way to repeatedly execute code inside a function through the function calling itself.
Functions that call themselves are know as _recursive_ functions.
Recursion can be viewed as another way to loop/iterate. 
And like looping, a Boolean expression or `True/False` test is used to know when to stop the recursive execution.
_Unlike_ looping, recursion without termination in Python cannot not run infinitely.
Values used in each function call are placed in their own frame on the Python interpreter stack.
If the total amount of function calls takes up more space than the stack has room for, it will result in an error.

## Looping vs Recursive Implementation

Looping and recursion may _feel_ similar in that they are both iterative.
However, they both _look_ differently, both at the code level and at the implementation level.
Looping can take place within the same frame on the call stack.
It usually manages this by updating one or more variables to maintain state as it iterates.

> This is an efficient implementation, but it can be somewhat cluttered at the code level.

Author: I think I'll change it to 

```
This is an efficient implementation, but it can be somewhat cluttered when looking at the code.
```

> Recursion, instead of updating state variables, can pass updated values as arguments to the next call of the same function.

Reviewer: ```suggestion

Recursion, rather than updating _variable state_, can pass _updated values_ directly as arguments to the next call (iteration) of the same function.
```

> This declutters the body of the function, but it is a less effficient implementation, as each call to the same function adds another frame to the stack.

Reviewer: ```suggestion
This declutters the body of the function and can clarify how each update happens.
However,  it is also a less efficient implementation, as each call to the same function adds another frame to the stack.
```

## Recursion: Why and Why Not?

> If there is risk of causing a stack error or overflow, why would anyone use a recursive strategy to solve a problem?

Reviewer: ```suggestion
If there is an "efficiency penalty" and risk of causing a stack error or overflow, why would anyone use a recursive strategy to solve a problem?
```

> There may be situations where a solution is more readable and/or easier to reason through when expressed through recursion than when expressed through looping.

Reviewer: ```suggestion
_Readability, traceability, and intent._ 
There may be situations where a solution is more readable and/or easier to reason through when expressed through recursion than when expressed through looping.
```

There may also be program constraints with using/mutating data, managing complexity, delegating responsibility, or organizing workloads.

> Problems that lend themselves to recursion include complex but repetitive problems that grow smaller over time, particularly [divide and conquer][divide and conquer] algorithms and [cumulative][cumulative] algorithms.

Reviewer: ```suggestion
However, due to Python's limit for how many frames are allowed on the stack, not all problems will benefit from a fully recursive strategy.
Problems that lend themselves to recursion include complex but repetitive problems that grow smaller over time, particularly [divide and conquer][divide and conquer] algorithms and [cumulative][cumulative] algorithms.
```

Author: if I'm reading this correctly this wants to intersperse the negatives of recursion with the positive. I prefer covering the positive and then covering the negative, as in the original order.

> However, due to Python's limit for how many frames are allowed on the stack, not all problems will benefit from a fully recursive strategy.

Reviewer: ```suggestion
```

Author: This was your addition, and I thought it was a good one, so I don't want to delete it..

In the future, Adya may try to work through problems recursively first.
She may find it easier to initially walk through the problem in clear steps when nesting, mutation, and complexity are minimized.
After working out the basic logic, she can then focus on optimizing her initial recursive steps into a more performant looping approach.

Even later, when she learns about `tuples`, Adya could consider further "optimizing" approaches, such as using a `list comprehension` with `Calendar.itermonthdates`, or memoizing certain values.

## Recursive Variation: The Tail Call

A tail call is when the last statement of a function only calls itself and nothing more.
This example is not a tail call, as the function adds 1 to the result of calling itself

```python
def print_increment(step, max_value):
    if step > max_value:
        return 1
    print(f'The step is {step}')
    return 1 + print_increment(step + 1, max_value)

def main():
    retval = print_increment(1, 2)
    print(f'retval is {retval} after recursion')

if __name__ == "__main__":
    main()

```

This will print

```
The step is 1
The step is 2
retval is 3 after recursion
```

> To refactor it to a tail call, make `j` a parameter of `printIncrement`

Reviewer: ```suggestion
To refactor it to a tail call, make `retval` a parameter of `print_increment`
```

```python
def print_increment(step, max_value, retval):
    if step > max_value:
        return retval
    print(f'The step is {step}')
    return print_increment(step + 1, max_value, retval + 1)

def main():
    retval = print_increment(1, 2, 1)
    print(f'retval is {retval} after recursion')

if __name__ == "__main__":
    main()

```

You may find a tail call even easier to reason through than a recursive call that is not a tail call.
However, it is always important when using recursion to know that there will not be so many iterations that the stack will overflow.

## Recursion Limits in Python

> Some languages are able to optimize tail calls so that each recursive call reuses the stack frame of the first call to the function, instead of adding another stack frame.

Reviewer: ```suggestion
Some languages are able to optimize tail calls so that each recursive call reuses the stack frame of the first call to the function (_similar to the way a loop reuses a frame_), instead of adding an additional frame to the stack.
```

# About

Recursion is a way to repeat code in a function by the function calling itself.
It is another way to loop/iterate. Like looping, the function should test a Boolean expression to know when to end.
Unlike looping, recursion that never ends will not run infinitely.
The values used in each function call will be placed in their own frame on the stack.

> If the amount of function calls takes up more space than the stack has, it will result in a stack overflow error.

Reviewer: ```suggestion
If the total amount of function calls takes up more space than the stack has room for, it will result in an error.

Python's interpreter stack is backed up by a C stack that is connected to the underlying operating system.
Function calls that take up more memory space than the C stack has will cause a stack overflow.
A stack overflow can crash Python or even the host operating system, if it is severe enough.
To avoid this situation, the Python interpreter has a maximum stack depth that limits the number of frames that can be placed on its stack.
The default is typically 999 frames, but varies by system.
The recursion limit for a given version of Python can be found by calling [`sys.getrecursionlimit()`][getrecursionlimit]
```

> ```python
> def printIncrement(i, maxValue):
>     if i > maxValue:
>         return
>     print(f'The number is {i}')
>     printIncrement(i + 1, maxValue)
> 

Reviewer: ```suggestion
```python
def print_increment(step, max_value):
    if step > max_value:
        return
    print(f'The number is {step}')
    print_increment(step + 1, max_value)

```

Reviewer: So this is the obligatory "PEP8 and examples need to be exemplars" nag.  Sorry I need to do this.  The gist is:
1.  snake_case function names and variable names.
2.  no single letter variable names - not even in comprehensions.
3.  REPL formatting for examples, if at all possible (example below).

```python
>>> def print_increment(step, max_value):
...     if step > max_value:
...         return
...     print(f'The number is {step}')
...     print_increment(step + 1, max_value)

```

Not sure here and below how to make that work, exactly.  We may want to replace this section with diagrams on the same function looping vs recursion, or with a link to Python Tutor like [this one](https://pythontutor.com/visualize.html#code=def%20steps_recurse%28number,%20step%3D0%29%3A%0A%20%20%20%20%22%22%22Calculate%20number%20of%20steps%20to%20reach%201.%0A%20%20%20%20%0A%20%20%20%20%3Aparam%20number%3A%20int%20-%20number%20to%20run%20through%20the%20conjecture.%0A%20%20%20%20%3Aparam%20step%3A%20int%20-%20current%20step%20number/iteration.%0A%20%20%20%20%3Areturn%3A%20step%3A%20int%20-%20total%20number%20of%20steps%20taken%20to%20reach%201.%0A%20%20%20%20%22%22%22%0A%0A%20%20%20%20new_number%20%3D%20number//2%20if%20number%20%25%202%20%3D%3D%200%20else%20%28number%20*%203%29%20%2B%201%0A%20%20%20%20%0A%20%20%20%20return%20step%20if%20number%20%3D%3D%201%20else%20steps_recurse%28new_number,%20step%3Dstep%20%2B%201%29%0A%0A%0A%0Adef%20steps_loop%28number%29%3A%0A%20%20%20%20%22%22%22Calculate%20number%20of%20steps%20to%20reach%201.%0A%20%20%20%20%0A%20%20%20%20%3Aparam%20number%3A%20int%20-%20number%20to%20run%20through%20the%20conjecture.%0A%20%20%20%20%3Aparam%20step%3A%20int%20-%20current%20step%20number/iteration.%0A%20%20%20%20%3Areturn%3A%20step%3A%20int%20-%20total%20number%20of%20steps%20taken%20to%20reach%201.%0A%20%20%20%20%22%22%22%0A%0A%20%20%20%20step%20%3D%200%20%0A%20%20%20%20while%20number%20!%3D%201%3A%0A%20%20%20%20%20%20%20%20number%20%3D%20number//2%20if%20number%20%25%202%20%3D%3D%200%20else%20%28number%20*%203%29%20%2B%201%0A%20%20%20%20%20%20%20%20step%20%2B%3D%201%0A%20%20%20%20%0A%20%20%20%20return%20step%0A%0Asteps_recurse%285%29%0Asteps_loop%285%29%0A%0A&cumulative=false&heapPrimitives=nevernest&mode=edit&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false).

Author: Thanks! I ran the code through [pep8online](http://pep8online.com/) until there were no issues reported, but that was evidently not sufficient. I'll probably take Kotlin off my machine and put Python back on with a good linter.
